File: morphac/robot/blueprint/test/robot_test.cc
#include "Eigen/Dense"
#include "gtest/gtest.h"

#include "robot/blueprint/include/robot.h"

namespace {

using std::make_shared;
using std::shared_ptr;
using std::srand;

using Eigen::VectorXd;
using Eigen::MatrixXd;

using morphac::constructs::Input;
using morphac::constructs::Pose;
using morphac::constructs::State;
using morphac::constructs::Velocity;
using morphac::mechanics::models::KinematicModel;
using morphac::robot::blueprint::Footprint;
using morphac::robot::blueprint::Robot;

class CustomKinematicModel : public KinematicModel {
 public:
  CustomKinematicModel(int pose_size, int velocity_size, int size_input)
      : KinematicModel(pose_size, velocity_size, size_input) {}

  State ComputeStateDerivative(const State& state, const Input& input) const {
    // f(x, u) = x * u  - x
    VectorXd derivative_vector(state.get_size());
    derivative_vector << state.get_data();
    derivative_vector =
        (derivative_vector.array() * input.get_data().array() -
         derivative_vector.array())
            .matrix();

    State derivative = State::CreateLike(state);
    derivative.set_pose_data(derivative_vector.head(pose_size));
    derivative.set_velocity_data(derivative_vector.tail(velocity_size));

    return derivative;
  }
};

class RobotTest : public ::testing::Test {
 protected:
  RobotTest() {
    // Set random seed for Eigen.
    srand(7);

    pose_vector_ = VectorXd::Random(3);
    velocity_vector_ = VectorXd::Random(2);
    input_vector_ = VectorXd::Random(5);
    footprint_matrix_ = MatrixXd::Random(10, 2);
  }

  void SetUp() override {}

  VectorXd pose_vector_, velocity_vector_, input_vector_;
  MatrixXd footprint_matrix_;
};

TEST_F(RobotTest, Construction) {
  CustomKinematicModel model(3, 2, 5);
  Footprint footprint(footprint_matrix_);
  Robot robot1{model, footprint};

  // Constructing with a temporary KinematicModel.
  Robot robot2{CustomKinematicModel{3, 2, 5}, footprint, State(3, 2)};
}

TEST_F(RobotTest, InvalidConstruction) {
  CustomKinematicModel model(3, 2, 5);
  Footprint footprint(footprint_matrix_);

  // Test invalid state/input dimensions.
  ASSERT_THROW(Robot(model, footprint, State(3, 3)), std::invalid_argument);
  ASSERT_THROW(Robot(model, footprint, State(4, 2)), std::invalid_argument);
  ASSERT_THROW(Robot(model, footprint, State(4, 3)), std::invalid_argument);
}

TEST_F(RobotTest, Accessors) {
  CustomKinematicModel model(3, 2, 5);
  Footprint footprint(footprint_matrix_);
  // Robot with default id.
  Robot robot1{model, footprint};

  VectorXd initial_pose{VectorXd::Random(3)};
  VectorXd initial_velocity{VectorXd::Random(2)};
  VectorXd initial_state(5);
  initial_state << initial_pose, initial_velocity;
  Robot robot2{model, footprint, State(initial_pose, initial_velocity)};

  // Making sure that the initial state is correct.
  // For the robot object created without initial state, the initial pose and
  // velocity must be zero. For the case with an explicit initial state, it must
  // match the given initial state.
  ASSERT_TRUE(
      robot1.get_state().get_data().isApprox(VectorXd::Zero(5)));
  ASSERT_TRUE(robot1.get_pose().get_data().isApprox(VectorXd::Zero(3)));
  ASSERT_TRUE(
      robot1.get_velocity().get_data().isApprox(VectorXd::Zero(2)));

  ASSERT_TRUE(robot2.get_state().get_data().isApprox(initial_state));
  ASSERT_TRUE(robot2.get_state().get_pose_data().isApprox(initial_pose));
  ASSERT_TRUE(
      robot2.get_state().get_velocity_data().isApprox(initial_velocity));
}

TEST_F(RobotTest, SetState) {
  CustomKinematicModel model(3, 2, 5);
  Footprint footprint(footprint_matrix_);
  Robot robot1{model, footprint};

  VectorXd pose_vector = VectorXd::Random(3);
  VectorXd velocity_vector = VectorXd::Random(2);

  robot1.set_state(State(VectorXd::Ones(3), VectorXd::Ones(2)));
  ASSERT_TRUE(
      robot1.get_state().get_data().isApprox(VectorXd::Ones(5)));

  robot1.set_pose(Pose(pose_vector));
  ASSERT_TRUE(robot1.get_pose().get_data().isApprox(pose_vector));
  // Making sure that the velocity has not changed.
  ASSERT_TRUE(
      robot1.get_velocity().get_data().isApprox(VectorXd::Ones(2)));

  robot1.set_velocity(Velocity(velocity_vector));
  // Making sure that the pose has not changed.
  ASSERT_TRUE(robot1.get_pose().get_data().isApprox(pose_vector));
  ASSERT_TRUE(
      robot1.get_velocity().get_data().isApprox(velocity_vector));

  // Test invalid state, pose and velocity setting.
  ASSERT_THROW(robot1.set_state(State(4, 2)), std::invalid_argument);
  ASSERT_THROW(robot1.set_state(State(3, 3)), std::invalid_argument);
  ASSERT_THROW(robot1.set_state(State(4, 3)), std::invalid_argument);
  ASSERT_THROW(robot1.set_state(State(VectorXd::Zero(2), VectorXd::Zero(2))),
               std::invalid_argument);
  ASSERT_THROW(robot1.set_state(State(VectorXd::Zero(3), VectorXd::Zero(1))),
               std::invalid_argument);
  ASSERT_THROW(robot1.set_state(State(VectorXd::Zero(2), VectorXd::Zero(1))),
               std::invalid_argument);

  ASSERT_THROW(robot1.set_pose(Pose(2)), std::invalid_argument);
  ASSERT_THROW(robot1.set_pose(Pose(VectorXd::Zero(5))), std::invalid_argument);

  ASSERT_THROW(robot1.set_velocity(Velocity(1)), std::invalid_argument);
  ASSERT_THROW(robot1.set_velocity(Velocity(VectorXd::Zero(3))),
               std::invalid_argument);
}

TEST_F(RobotTest, Derivative) {
  CustomKinematicModel model(3, 2, 5);
  Footprint footprint(footprint_matrix_);
  VectorXd initial_pose(3), initial_velocity(2), input(5),
      expected_state_derivative1(5), expected_state_derivative2(5);

  initial_pose << 1, 1, -2;
  initial_velocity << 3, -5;
  input << 1, 2, 3, 2, 1;
  expected_state_derivative1 << 0, 1, -4, 3, 0;

  Robot robot{model, footprint, State(initial_pose, initial_velocity)};

  // Testing derivative computation.
  State derivative1 = robot.ComputeStateDerivative(input);

  ASSERT_TRUE(
      derivative1.get_data().isApprox(expected_state_derivative1));

  expected_state_derivative2 << 0, 1, 2, 1, 0;

  // Testing derivative computation for a different state.
  State derivative2 = robot.ComputeStateDerivative(
      State(VectorXd::Ones(3), VectorXd::Ones(2)), input);

  ASSERT_TRUE(
      derivative2.get_data().isApprox(expected_state_derivative2));
}

}  // namespace

int main(int argc, char** argv) {
  ::testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}

File: morphac/robot/blueprint/python/test/robot_test.py
import numpy as np
import pytest

from morphac.constructs import Pose, ControlInput, State, Velocity
from morphac.mechanics.models import KinematicModel, DiffDriveModel, TricycleModel
from morphac.robot.blueprint import Footprint, Robot

# KinematicModel extension class to ensure that a robot can be created from
# a subclassed KinematicModel


class CustomKinematicModel(KinematicModel):
    def __init__(self, sp, sv, si, a, b):

        KinematicModel.__init__(self, sp, sv, si)
        self.a = a
        self.b = b

    def compute_state_derivative(self, robot_state, control_input):

        tmp_der = np.sum(np.multiply(robot_state.data, control_input.data))
        tmp_der = tmp_der + self.a * self.b
        # tmp_der is a scalar, but the derivative must be a State object.
        # Copying tmp_der to each State element.
        der = State([tmp_der] * self.pose_size, [tmp_der] * self.velocity_size)
        return der


@pytest.fixture()
def generate_robot_list():

    r1 = Robot(DiffDriveModel(1, 1), Footprint([[1, 2]]))
    r2 = Robot(TricycleModel(1, 1), Footprint(np.ones([20, 2])))
    r3 = Robot(CustomKinematicModel(3, 2, 5, 2.5, 2), Footprint([[1, 0], [0, 1]]))
    r4 = Robot(
        kinematic_model=CustomKinematicModel(2, 2, 4, 0, 0),
        footprint=Footprint([[0, 0]]),
        initial_state=State([1, 2], [3, 4]),
    )

    return r1, r2, r3, r4


def test_invalid_construction():

    with pytest.raises(ValueError):
        # Invalid State dimensions.
        _ = Robot(DiffDriveModel(1, 1), Footprint([[0, 0]]), State(1, 1))


def test_kinematic_model(generate_robot_list):

    r1, r2, r3, r4 = generate_robot_list

    # Making sure that they are of the right type.
    assert isinstance(r1.kinematic_model, DiffDriveModel)
    assert isinstance(r2.kinematic_model, TricycleModel)
    assert isinstance(r3.kinematic_model, CustomKinematicModel)
    assert isinstance(r4.kinematic_model, CustomKinematicModel)

    # Testing the actual model details.
    assert r1.kinematic_model.pose_size == 3
    assert r2.kinematic_model.pose_size == 4
    assert r3.kinematic_model.pose_size == 3
    assert r4.kinematic_model.pose_size == 2

    assert r1.kinematic_model.velocity_size == 0
    assert r2.kinematic_model.velocity_size == 0
    assert r3.kinematic_model.velocity_size == 2
    assert r4.kinematic_model.velocity_size == 2

    assert r1.kinematic_model.control_input_size == 2
    assert r2.kinematic_model.control_input_size == 2
    assert r3.kinematic_model.control_input_size == 5
    assert r4.kinematic_model.control_input_size == 4

    assert r1.kinematic_model.radius == 1
    assert r1.kinematic_model.width == 1
    assert r2.kinematic_model.width == 1
    assert r2.kinematic_model.length == 1
    assert r3.kinematic_model.a == 2.5
    assert r3.kinematic_model.b == 2
    assert r4.kinematic_model.a == 0
    assert r4.kinematic_model.b == 0

    # Making sure that the kinematic model is read only.
    with pytest.raises(AttributeError):
        r1.kinematic_model = TricycleModel(2, 2)
    with pytest.raises(AttributeError):
        r2.kinematic_model = TricycleModel(1, 1)


def test_footprint(generate_robot_list):

    r1, r2, r3, r4 = generate_robot_list

    assert np.allclose(r1.footprint.data, [[1, 2]])
    assert np.allclose(r2.footprint.data, np.ones([20, 2]))
    assert np.allclose(r3.footprint.data, [[1, 0], [0, 1]])
    assert np.allclose(r4.footprint.data, [[0, 0]])

    # Making sure that footprint is read only.
    with pytest.raises(AttributeError):
        r1.footprint = [[1, 2], [3, 4], [5, 6]]
    with pytest.raises(AttributeError):
        r2.footprint = np.ones([20, 2])


def test_state(generate_robot_list):

    r1, r2, r3, r4 = generate_robot_list

    # r1, r2, r3 have not been created using an initial state and must hence
    # be zero.
    assert np.allclose(r1.state.data, [0, 0, 0])
    assert np.allclose(r2.state.data, [0, 0, 0, 0])
    assert np.allclose(r3.state.data, [0, 0, 0, 0, 0])
    # r4 has an initial state.
    assert np.allclose(r4.state.data, [1, 2, 3, 4])

    # Test state setting.
    r1.state = State([1, 1, 1], [])
    r3.state.data = [1, 1, 2, 3, 5]

    assert np.allclose(r1.state.data, [1, 1, 1])
    assert np.allclose(r3.state.data, [1, 1, 2, 3, 5])

    # State operations.
    r1.state -= State([0, 1, 0], [])
    r3.state = r3.state + State([1, 1, 1], [1, 1])
    assert np.allclose(r1.state.data, [1, 0, 1])
    assert np.allclose(r3.state.data, [2, 2, 3, 4, 6])

    r1.state *= 3
    r3.state = -10 * r3.state
    assert np.allclose(r1.state.data, [3, 0, 3])
    assert np.allclose(r3.state.data, [-20, -20, -30, -40, -60])


def test_pose(generate_robot_list):

    r1, r2, r3, r4 = generate_robot_list

    assert np.allclose(r1.pose.data, [0, 0, 0])
    assert np.allclose(r2.pose.data, [0, 0, 0, 0])
    assert np.allclose(r3.pose.data, [0, 0, 0])
    assert np.allclose(r4.pose.data, [1, 2])

    # Test pose setting.
    r1.pose = Pose([2, 2, 2])
    r3.pose = Pose([1, 1, 1])

    # Also making sure that the velocity component has not changed.
    assert np.allclose(r1.pose.data, [2, 2, 2])
    assert np.allclose(r3.pose.data, [1, 1, 1])
    assert np.allclose(r1.state.data, [2, 2, 2])
    assert np.allclose(r3.state.data, [1, 1, 1, 0, 0])

    # Pose operations.
    r1.pose -= Pose([1, 1, 1])
    r3.pose = r3.pose + Pose([1, 1, 1])
    assert np.allclose(r1.pose.data, [1, 1, 1])
    assert np.allclose(r3.pose.data, [2, 2, 2])
    assert np.allclose(r1.state.data, [1, 1, 1])
    assert np.allclose(r3.state.data, [2, 2, 2, 0, 0])

    r1.pose *= 2
    r3.pose = -1 * r3.pose
    assert np.allclose(r1.pose.data, [2, 2, 2])
    assert np.allclose(r3.pose.data, [-2, -2, -2])
    assert np.allclose(r1.state.data, [2, 2, 2])
    assert np.allclose(r3.state.data, [-2, -2, -2, 0, 0])


def test_velocity(generate_robot_list):

    r1, r2, r3, r4 = generate_robot_list

    assert np.allclose(r3.velocity.data, [0, 0])
    assert np.allclose(r4.velocity.data, [3, 4])

    # Test velocity setting.
    r3.velocity = Velocity([1, 1])
    r4.velocity = Velocity([0, 0])

    # Also making sure that the pose component has not changed.
    assert np.allclose(r3.velocity.data, [1, 1])
    assert np.allclose(r4.velocity.data, [0, 0])
    assert np.allclose(r3.state.data, [0, 0, 0, 1, 1])
    assert np.allclose(r4.state.data, [1, 2, 0, 0])

    # Velocity operations.
    r3.velocity -= Velocity([2, 2])
    r4.velocity = r4.velocity + Velocity([2, 2])
    assert np.allclose(r3.velocity.data, [-1, -1])
    assert np.allclose(r4.velocity.data, [2, 2])
    assert np.allclose(r3.state.data, [0, 0, 0, -1, -1])
    assert np.allclose(r4.state.data, [1, 2, 2, 2])

    r3.velocity *= 2
    r4.velocity = -5 * r4.velocity
    assert np.allclose(r3.velocity.data, [-2, -2])
    assert np.allclose(r4.velocity.data, [-10, -10])
    assert np.allclose(r3.state.data, [0, 0, 0, -2, -2])
    assert np.allclose(r4.state.data, [1, 2, -10, -10])


def test_derivative_computation(generate_robot_list):

    r1, r2, r3, r4 = generate_robot_list

    # We check for the actual derivative values only for r3 and r4 as they
    # have the custom kinematic model. The other models have already been
    # tested on the cpp side.
    der1 = r1.compute_state_derivative(ControlInput([0, 0]))
    der2 = r2.compute_state_derivative(State([1, 1, 1, 1], []), ControlInput([0, 0]))

    # Test with positional arguments.
    der3 = r3.compute_state_derivative(
        robot_state=State([1, 0, 0], [-1, 0]),
        control_input=ControlInput([2, 2, 2, -2, 2]),
    )
    der4 = r4.compute_state_derivative(control_input=ControlInput([-1, 1, -2, 2]))

    assert np.allclose(der1.data, [0, 0, 0])
    assert np.allclose(der2.data, [0, 0, 0, 0])
    assert np.allclose(der3.data, [9] * 5)
    assert np.allclose(der4.data, [3] * 4)

    # Making sure that the derivative computation cannot be called with
    # incorrect state/control input dimensions, even with a custom kinematic model.
    with pytest.raises(ValueError):
        r1.compute_state_derivative(ControlInput([1, 1, 1]))
    with pytest.raises(ValueError):
        r2.compute_state_derivative(State(3, 0), ControlInput(2))
    with pytest.raises(ValueError):
        r3.compute_state_derivative(ControlInput(3))
    with pytest.raises(ValueError):
        r4.compute_state_derivative(State([1, 1], [2, 2]), ControlInput(5))
    with pytest.raises(ValueError):
        r4.compute_state_derivative(State([1, 1, 0], [2, 2]), ControlInput(2))
